src/modules/logger/watchdog.cpp

```diff
@@ -0,0 +1,171 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "watchdog.h"
+
+#include <px4_log.h>
+
+using namespace time_literals;
+
+namespace px4
+{
+namespace logger
+{
+
+bool watchdog_update(watchdog_data_t &watchdog_data)
+{
+
+#ifdef __PX4_NUTTX
+
+	if (system_load.initialized && watchdog_data.logger_main_task_index >= 0
+	    && watchdog_data.logger_writer_task_index >= 0) {
+		const hrt_abstime now = hrt_absolute_time();
+		const system_load_taskinfo_s &log_writer_task = system_load.tasks[watchdog_data.logger_writer_task_index];
+
+		if (log_writer_task.valid) {
+			// Trigger the watchdog if the log writer task has been ready to run for a
+			// minimum duration and it has not been scheduled during that time.
+			// When the writer is waiting for an SD transfer, it is not in ready state, thus a long dropout
+			// will not trigger it. The longest period in ready state I measured was around 70ms,
+			// after a param change.
+			// We only check the log writer because it runs at lower priority than the main thread.
+			// No need to lock the tcb access, since we are in IRQ context
+
+			// update the timestamp if it has been scheduled recently
+			if (log_writer_task.curr_start_time > watchdog_data.ready_to_run_timestamp) {
+				watchdog_data.ready_to_run_timestamp = log_writer_task.curr_start_time;
+			}
+
+			// update the timestamp if not ready to run or if transitioned into ready to run
+			uint8_t current_state = log_writer_task.tcb->task_state;
+
+			if (current_state != TSTATE_TASK_READYTORUN
+			    || (watchdog_data.last_state != TSTATE_TASK_READYTORUN && current_state == TSTATE_TASK_READYTORUN)) {
+				watchdog_data.ready_to_run_timestamp = now;
+			}
+
+			watchdog_data.last_state = current_state;
+
+#if 0 // for debugging
+			// test code that prints the maximum time in ready state
+			static uint64_t max_time = 0;
+
+			if (now - watchdog_data.ready_to_run_timestamp > max_time) {
+				max_time = now - watchdog_data.ready_to_run_timestamp;
+			}
+
+			static int counter = 0;
+
+			if (++counter > 300) {
+				PX4_ERR("max time in ready: %i ms", (int)max_time / 1000);
```

Comment:
Is printf called off of this in interrupt context?

Reply:
Yes. I thought this was safe?

Reply:
My recollection is that there is a possibility of a  [wait](https://bitbucket.org/nuttx/nuttx/src/21120057e893b53c128c2a5baeb5a0ee3615a186/sched/semaphore/sem_wait.c#lines-94)on semaphore in the code path of PX4_ERR.  Is that not the case?

Reply:
I'll check

Reply:
You are right, `printf` calls  `lib_sem_initialize` which uses a semaphore. What are you using for such cases?
In any case, thanks for the hint. I removed them now.